### PR TITLE
Modify pointwise cat heuristic to only apply when inputs are all pointwise and outputs are all pointwise

### LINKED ISSUE #112139: [inductor][cpu] performance regression
<p>new_perf_regression in 2023-10-22</p>
<p>Note: multi threads secnario for models above first *, single thread for models between two *</p>
<table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>name</th>
      <th>batch_size_new</th>
      <th>speed_up_new</th>
      <th>inductor_new</th>
      <th>eager_new</th>
      <th>compilation_latency_new</th>
      <th>batch_size_old</th>
      <th>speed_up_old</th>
      <th>inductor_old</th>
      <th>eager_old</th>
      <th>compilation_latency_old</th>
      <th>Ratio Speedup(New/old)</th>
      <th>Eager Ratio(old/new)</th>
      <th>Inductor Ratio(old/new)</th>
      <th>Compilation_latency_Ratio(old/new)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>basic_gnn_gin</td>
      <td>1</td>
      <td>1.108124</td>
      <td>0.005866914</td>
      <td>0.006501268209336001</td>
      <td>14.090111</td>
      <td>1</td>
      <td>1.123217</td>
      <td>0.005190271</td>
      <td>0.0058298006218069996</td>
      <td>14.043207</td>
      <td>0.99</td>
      <td>0.9</td>
      <td>0.88</td>
      <td>1.0</td>
    </tr>
    <tr>
      <td>cm3leon_generate</td>
      <td>1</td>
      <td>0.497383</td>
      <td>3.907282571</td>
      <td>1.943415927011693</td>
      <td>373.98615</td>
      <td>1</td>
      <td>0.616159</td>
      <td>3.171211898</td>
      <td>1.9539707518597822</td>
      <td>453.802624</td>
      <td>0.81</td>
      <td>1.01</td>
      <td>0.81</td>
      <td>1.21</td>
    </tr>
    <tr>
      <td>hf_Whisper</td>
      <td>8</td>
      <td>1.408156</td>
      <td>0.126020063</td>
      <td>0.17745590783382797</td>
      <td>29.196827</td>
      <td>8</td>
      <td>1.703685</td>
      <td>0.103420151</td>
      <td>0.17619535995643498</td>
      <td>29.233589</td>
      <td>0.83</td>
      <td>0.99</td>
      <td>0.82</td>
      <td>1.0</td>
    </tr>
    <tr>
      <td>phlippe_densenet</td>
      <td>128</td>
      <td>1.558358</td>
      <td>0.02617806</td>
      <td>0.040794789225479994</td>
      <td>55.679105</td>
      <td>128</td>
      <td>2.048464</td>
      <td>0.01980409</td>
      <td>0.04056796541776</td>
      <td>55.710849</td>
      <td>0.76</td>
      <td>0.99</td>
      <td>0.76</td>
      <td>1.0</td>
    </tr>
    <tr>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
    </tr>
    <tr>
      <td>basic_gnn_edgecnn</td>
      <td>1</td>
      <td>0.871592</td>
      <td>0.37231181199999996</td>
      <td>0.324503996844704</td>
      <td>11.772577</td>
      <td>1</td>
      <td>1.330474</td>
      <td>0.24396664999999998</td>
      <td>0.32459128469209997</td>
      <td>11.120962</td>
      <td>0.66</td>
      <td>1.0</td>
      <td>0.66</td>
      <td>0.94</td>
    </tr>
    <tr>
      <td>phlippe_densenet</td>
      <td>1</td>
      <td>0.900174</td>
      <td>0.005640694</td>
      <td>0.005077606080756</td>
      <td>54.921562</td>
      <td>1</td>
      <td>1.328827</td>
      <td>0.003758348</td>
      <td>0.004994194297796</td>
      <td>54.409835</td>
      <td>0.68</td>
      <td>0.98</td>
      <td>0.67</td>
      <td>0.99</td>
    </tr>
    <tr>
      <td>mixnet_l</td>
      <td>1</td>
      <td>0.736023</td>
      <td>0.050399831</td>
      <td>0.037095434812113</td>
      <td>91.170201</td>
      <td>1</td>
      <td>1.148282</td>
      <td>0.032940426</td>
      <td>0.037824898248132004</td>
      <td>92.450105</td>
      <td>0.64</td>
      <td>1.02</td>
      <td>0.65</td>
      <td>1.01</td>
    </tr>
    <tr>
      <td>tf_mixnet_l</td>
      <td>1</td>
      <td>1.088825</td>
      <td>0.036983614000000005</td>
      <td>0.04026868351355</td>
      <td>111.028423</td>
      <td>1</td>
      <td>1.275501</td>
      <td>0.031524481</td>
      <td>0.040209507039981</td>
      <td>112.353437</td>
      <td>0.85</td>
      <td>1.0</td>
      <td>0.85</td>
      <td>1.01</td>
    </tr>
    <tr>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
      <td>*</td>
    </tr>
  </tbody>
</table>
<p>SW info</p>
<table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>SW</th>
      <th>Nightly commit</th>
      <th>Main commit</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Pytorch</td>
      <td>f468e74</td>
      <td>bf01a7b</td>
    </tr>
    <tr>
      <td>Torchbench</td>
      <td>/</td>
      <td>7617d3f5</td>
    </tr>
    <tr>
      <td>torchaudio</td>
      <td>475b6ae</td>
      <td>ede4309</td>
    </tr>
    <tr>
      <td>torchtext</td>
      <td>142d029</td>
      <td>45e4b8c</td>
    </tr>
    <tr>
      <td>torchvision</td>
      <td>8636bf3</td>
      <td>4ac707a</td>
    </tr>
    <tr>
      <td>torchdata</td>
      <td>eb9bf61</td>
      <td>d76d92c</td>
    </tr>
    <tr>
      <td>dynamo_benchmarks</td>
      <td>0200b11</td>
      <td>/</td>
    </tr>
  </tbody>
</table>
<p>Reference SW info (nightly)</p>
<table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>item</th>
      <th>commit</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>7617d3f5</td>
    </tr>
    <tr>
      <td>torch</td>
      <td>2.2.0a0+git8aa4019</td>
    </tr>
    <tr>
      <td>torchvision</td>
      <td>0.16.0a0+8636bf3</td>
    </tr>
    <tr>
      <td>torchtext</td>
      <td>0.16.0a0+142d029</td>
    </tr>
    <tr>
      <td>torchaudio</td>
      <td>2.2.0a0+475b6ae</td>
    </tr>
    <tr>
      <td>torchdata</td>
      <td>0.7.0a0+eb9bf61</td>
    </tr>
    <tr>
      <td>dynamo_benchmarks</td>
      <td>0200b11</td>
    </tr>
  </tbody>
</table>
<p>image: docker pull ccr-registry.caas.intel.com/pytorch/pt_inductor:2023_10_23_aws</p>

<p>Repro</p>
<a href=https://github.com/chuanqi129/inductor-tools/blob/yudong/aws_auto/scripts/modelbench/inductor_single_run.sh>
  inductor_single_run.sh </a>
<code>
bash inductor_single_run.sh multiple/single inference performance torchbench/huggingface/timm_models model float32 first static default 0
</code>

<p><a href=>
  Suspected guilty commit </a></p>

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114984
* __->__ #114520

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -1074,9 +1074,8 @@ def should_lower_cat_input(x) -> bool:
     if len(inputs) <= config.max_pointwise_cat_inputs:
         pointwise_uses = all(is_pointwise_use(use) for use in V.current_node.users)
         all_pointwise_inputs = all(should_lower_cat_input(inp) for inp in inputs)
-        any_pointwise_inputs = any(should_lower_cat_input(inp) for inp in inputs)
 
-        if all_pointwise_inputs or (any_pointwise_inputs and pointwise_uses):
+        if all_pointwise_inputs and pointwise_uses:
```
Comment (eellison) [inline torch/_inductor/lowering.py:1078]: I think we should check if the pointwise input will get fused into the reduction, instead of this change. if you had `out = torch.cat([torch.softmax(a, dim=-1) + 1, torch.softmax(b, dim=-1) + 1])` this would fuse

Comment (Chillee) [reply torch/_inductor/lowering.py:1078]: I think it's not that easy for us at this point to predict whether the pointwise will get fused into a reduction.

Nevertheless, if there are no downstream pointwise ops to fuse into, then this is not beneficial no? So given that, I think we should err on the side of caution.

Comment (eellison) [reply torch/_inductor/lowering.py:1078]: I think we should just get rid of the `all_pointwise_inputs` heuristic. If there are `any_pointwise_inputs and pointwise_uses` then this saved global memory reads. 

If we need a more thorough fix for not fusing in reductions, I can work on that.

Comment (Chillee) [reply torch/_inductor/lowering.py:1078]: Oh, that's right. I'll change the heuristic.

Comment (eellison) [reply torch/_inductor/lowering.py:1078]: While you're add it, could you skip pointwise cat for cpu? There was regression bc of it, see : https://github.com/pytorch/pytorch/issues/112139.

**Review by eellison — APPROVED** [review]
